text/0078-improve-tso-scalability.md

```diff
@@ -0,0 +1,93 @@
+# Improve the Scalability of TSO Service
+
+- RFC PR: https://github.com/tikv/rfcs/pull/78
+- Tracking Issue: https://github.com/tikv/pd/issues/3149
+
+## Summary
+
+As the size of the cluster grows, some users encountered a situation that is
+extremely challenging for TSO performance. In this kind of cluster that may
+have many TiDB instances, a certain number of PD clients will request TSO to
+the PD leader concurrently, and this puts a lot of CPU pressure on the PD leader
+because of the Go Runtime scheduling caused by the gRPC connection switch. With
+the high CPU pressure, TSO requests will suffer from increasing average latency
+and long-tail latency, which will hurt TiDB's QPS performance. To improve the
+scalability of the TSO Service, we will propose two kinds of enhancement to both TSO
+client and server in this RFC to reduce the PD CPU usage and improve QPS performance.
+
+## Motivation
+
+As mentioned before, improvement is needed in the cluster that has a certain number
+of TiDB instances, i.e PD clients, to reduce the CPU pressure of PD leader. With
+better TSO handling performance, in this case, TSO service won't be the potential
+bottleneck of a big cluster easily and have better scalability.
+
+## Detailed design
+
+### Current TSO processing flow
+
+Before proposing specific improvements, let me give a brief overview of the current
+PD client and TSO processing flow first.
+
+For the PD client, every time it receive a TSO request from the upper level, it won't
+send the TSO gRPC request to the PD leader immediately. Instead, it will collect as
+many TSO requests as it can from a channel and batch them as a single request to
+send to reduce the gRPC request number it actually sent. For example, at a certain
+moment, PD client may receive 10 TSO requests at once, it will send just one TSO gRPC
+request to the PD leader with an argument `count` inside the gRPC request body.
+
+![TSO Client Batch - Old](../media/tso-client-batch-old.png)
+
+For the PD server, it just takes the request and returns an incremental and unique TSO
+request with 4 counts and return it to the client. After the PD client receives the
+response, it will split the single TSO request into 10 different TSOs and return it to
+the upper requester.
+
+During the whole process flow, every PD client will maintain a gRPC stream with the PD
+leader to send and receive the TSO request efficiently.
+
+### Enhancement #1: Improve the batch effect
+
+An intuitive improvement is to improve the batch effect of TSO, i.e. to reduce the number
+of TSO gRPC requests by increasing the size of each batch with the same number of TSO requests.
+With fewer TSO gRPC requests, the CPU pressure on the PD leader can be reduced directly.
+
+We can introduce different batch strategies such as waiting for a while to fetch more TSO requests
```

Comment:
A different way to improve the efficiency is enhance batching on PD side.

I'm not sure if grpc-go has [buffer_hint](https://grpc.github.io/grpc/cpp/md_doc_cpp_perf_notes.html) that can improve throughput and reduce CPU usage a lot, but it may help to return multiple response at the same time. You can check how it's implemented in mini-pd:

https://github.com/BusyJay/mini-pd/blob/148202ee0daf24adcae97786ac77ad7c0a05cbb9/src/net/service/pd.rs#L206-L281